runtime/src/iree/base/loop_emscripten.c

```diff
@@ -59,10 +83,121 @@ IREE_API_EXPORT void iree_loop_emscripten_free(iree_loop_emscripten_t* loop) {
+    return iree_make_status(IREE_STATUS_UNIMPLEMENTED,
+                            "wait handle export/import not implemented");
+  }
+
+  if (wait_handle.type != IREE_WAIT_PRIMITIVE_TYPE_PROMISE) {
+    return iree_make_status(IREE_STATUS_UNIMPLEMENTED,
+                            "only Promise wait primitives are supported");
+  }
+
+  *out_handle = wait_handle.value.promise.handle;
+  return iree_ok_status();
+}
+
+static iree_status_t iree_loop_emscripten_run_wait_one(
+    iree_loop_emscripten_t* loop_emscripten,
+    iree_loop_wait_one_params_t* params) {
+  int promise_handle = 0;
+  IREE_RETURN_IF_ERROR(iree_loop_emscripten_get_promise_handle(
+      params->wait_source, &promise_handle));
+  iree_loop_t loop = iree_loop_emscripten(loop_emscripten);
+  uint32_t timeout_ms =
+      iree_absolute_deadline_to_timeout_ms(params->deadline_ns);
+  return iree_loop_command_wait_one(loop_emscripten->scope, params->callback.fn,
+                                    params->callback.user_data, timeout_ms,
+                                    promise_handle, loop);
+}
+
+static iree_status_t iree_loop_emscripten_run_wait_any(
+    iree_loop_emscripten_t* loop_emscripten,
+    iree_loop_wait_multi_params_t* params) {
+  int* promise_handles = NULL;
+  IREE_RETURN_IF_ERROR(iree_allocator_malloc(loop_emscripten->allocator,
+                                             sizeof(int) * params->count,
+                                             (void**)&promise_handles));
+
+  iree_status_t status = iree_ok_status();
+  for (iree_host_size_t i = 0; i < params->count; ++i) {
+    if (iree_status_is_ok(status)) {
+      status = iree_loop_emscripten_get_promise_handle(params->wait_sources[i],
+                                                       &promise_handles[i]);
+    }
+  }
+
+  if (iree_status_is_ok(status)) {
+    iree_loop_t loop = iree_loop_emscripten(loop_emscripten);
+    uint32_t timeout_ms =
+        iree_absolute_deadline_to_timeout_ms(params->deadline_ns);
+    status = iree_loop_command_wait_any(
+        loop_emscripten->scope, params->callback.fn, params->callback.user_data,
+        timeout_ms, params->count, promise_handles, loop);
+  }
+
+  iree_allocator_free(loop_emscripten->allocator, promise_handles);
+  return status;
+}
+
+static iree_status_t iree_loop_emscripten_run_wait_all(
+    iree_loop_emscripten_t* loop_emscripten,
+    iree_loop_wait_multi_params_t* params) {
+  int* promise_handles = NULL;
+  IREE_RETURN_IF_ERROR(iree_allocator_malloc(loop_emscripten->allocator,
+                                             sizeof(int) * params->count,
+                                             (void**)&promise_handles));
+
+  iree_status_t status = iree_ok_status();
+  for (iree_host_size_t i = 0; i < params->count; ++i) {
+    if (iree_status_is_ok(status)) {
+      status = iree_loop_emscripten_get_promise_handle(params->wait_sources[i],
+                                                       &promise_handles[i]);
+    }
```

Comment:
No use looping once the status is set non-OK.
```suggestion
    status = iree_loop_emscripten_get_promise_handle(params->wait_sources[i],
                                                     &promise_handles[i]);
    if (!iree_status_is_ok(status)) break;
```